drake/solvers/Constraint.h

```diff
@@ -193,6 +195,47 @@ class BoundingBoxConstraint : public LinearConstraint {
     y = x;
   }
 };
+
+
+/** LinearComplementarityConstraint
+ *
+ * Implements a constraint of the form:
+ *   Mx + q >= 0
+ *   x >= 0
+ *   x'(Mx + q) == 0
+ *
+ * An implied slack variable complements any 0 component of x, which
+ * (TODO ggould) will in the future be retrievable.
+ */
+class LinearComplementarityConstraint : public Constraint {
+ public:
+  template <typename DerivedM, typename Derivedq>
+  LinearComplementarityConstraint(const Eigen::MatrixBase<DerivedM>& M,
+                                  const Eigen::MatrixBase<Derivedq>& q)
+      : Constraint(q.rows()), M(M), q(q) {}
+
+  virtual ~LinearComplementarityConstraint() {}
+
+  virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x,
+                    Eigen::VectorXd& y) const override {
+    throw std::runtime_error(
+        "Linear Complementarity Constraints should not invoke eval()");
```

Comment:
we will likely want to revisit this as some point... it could output a boolean if nothing else, which might be useful when we think about the integer relaxations of this.  but totally fine for now.